src/main.cpp

```diff
@@ -7254,13 +7261,20 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
 
     else if (strCommand == "getdata")
     {
-        vector<CInv> vInv;
-        vRecv >> vInv;
-        if (vInv.size() > MAX_INV_SZ)
+        auto vRecvStreamSz = ReadCompactSize(vRecv);
+        if (vRecvStreamSz > MAX_INV_SZ)
         {
             Misbehaving(pfrom->GetId(), 20);
-            return error("message getdata size() = %u", vInv.size());
+            return error("message getdata size() = %u", vRecvStreamSz);
+        }
+        else if(vRecvStreamSz * sizeof(CInv)!= vRecv.in_avail()) {
+            Misbehaving(pfrom->GetId(), 20);
+            return error("invalid message payload for %u", vRecvStreamSz);
         }
+        vRecv.Rewind(GetSizeOfCompactSize(vRecvStreamSz));
+
+        vector<CInv> vInv;
+        vRecv >> vInv;
```

Comment:
I like these checks before deserializing the payload.

I see the new code is currently being used in two places (`inv` and `getdata`), and it could be useful somewhere else, would it be possible to define a generic template function to reduce duplication?

In my mind, such a function should:
- Get the stream
- Check size
- Deserialize the stream and return the related data structure (or error)

Reply:
done